File: CivicSurvival.PublicAudit/FlatbuffersReader.cs
// Hand-rolled FlatBuffers reader used by the public-audit runner.
//
// Supports only what the audit needs:
//   - 4-byte file_identifier lock (CSWP)
//   - root uoffset (must be safe to follow)
//   - Envelope: payload_type (uint8), payload (union table offset)
//   - CommandBatch: schema_version (uint16), commands (vector of tables)
//   - CommandEnvelope: kind (uint8), payload (vector of bytes)
//   - ProjectionDelta: campaign_id/observer_id/state_hash/views/...
//     byte vectors; base_revision/new_revision/tick/validated_revision
//     uint32; decisions[] (vector of CommandDecision tables)
//
// Mirrors the wire layout documented in Google FlatBuffers (Apache 2.0)
// without pulling in the library. Bounds-checks every offset to reject
// corruption, truncation, and arbitrary-length attacks.
//
// Cross-validated against `flatc --binary` output via `flatc --json
// --raw-binary --strict-json` of the committed golden fixtures at
// .agileplus/civic-warfare-program/contracts/fixtures/.

using System.Buffers.Binary;

public static class FlatbuffersReader
{
    public enum RootPayloadKind : byte { NONE = 0, CommandBatch = 1, ProjectionDelta = 2, SaveEnvelope = 3 }

    public readonly record struct CommandView(int KindByte, int PayloadByteCount);

    public readonly record struct DecisionView(int CodeByte, int ReasonKeyByteCount, int DetailsByteCount);

    public sealed class EnvelopeView
    {
        public RootPayloadKind PayloadType { get; set; }
        public ushort SchemaVersion { get; set; }
        public int CommandCount => Commands.Count;
        public List<CommandView> Commands { get; } = new();

        // ProjectionDelta fields (populated when PayloadType == ProjectionDelta).
        public uint BaseRevision { get; set; }
        public uint NewRevision { get; set; }
        public uint Tick { get; set; }
        public int CampaignIdByteCount { get; set; }
        public int ObserverIdByteCount { get; set; }
        public int StateHashByteCount { get; set; }
        public int ViewsByteCount { get; set; }
        public int RemovalsByteCount { get; set; }
        public int AlertsByteCount { get; set; }
        public int ExplanationsByteCount { get; set; }
        public List<DecisionView> Decisions { get; } = new();
        public int DecisionCount => Decisions.Count;
    }

    public static bool TryParseEnvelope(byte[] buffer, string expectedFileIdentifier,
        string expectedRootType, out EnvelopeView envelope, out string? error)
    {
        envelope = null!;
        error = null;
        if (buffer is null || buffer.Length < 8) { error = "buffer < 8 bytes"; return false; }

        var span = buffer.AsSpan();
        var rootOffset = BinaryPrimitives.ReadUInt32LittleEndian(span[0..4]);
        if (rootOffset > buffer.Length - 4) { error = $"root offset {rootOffset} past EOF"; return false; }

        var ident = System.Text.Encoding.ASCII.GetString(span[4..8]);
        if (!string.Equals(ident, expectedFileIdentifier, StringComparison.Ordinal))
        { error = $"file_identifier {ident} != {expectedFileIdentifier}"; return false; }
        _ = expectedRootType;

        var rootTableOff = (int)rootOffset;
        var env = new EnvelopeView();
        if (!TryFollowTable(span, rootTableOff, out var envFields, out error)) return false;
        if (!TryReadUInt8Field(span, rootTableOff, envFields, 0, out var payloadTypeByte, out error))
            return false;
        env.PayloadType = (RootPayloadKind)payloadTypeByte;

        if (!TryReadUOffsetFieldLocation(span, rootTableOff, envFields, 1, out var payloadStoreAbs,
                out var payloadRel, out error)) return false;
        var payloadAbs = (int)(payloadStoreAbs + payloadRel);

        switch (env.PayloadType)
        {
            case RootPayloadKind.CommandBatch:
                if (!DecodeCommandBatch(span, payloadAbs, env, out error)) return false;
                break;
            case RootPayloadKind.ProjectionDelta:
                if (!DecodeProjectionDelta(span, payloadAbs, env, out error)) return false;
                break;
            default:
                // SaveEnvelope and NONE: leave EnvelopeView at defaults.
                break;
        }

        envelope = env;
        return true;
    }

    private static bool DecodeCommandBatch(ReadOnlySpan<byte> span, int payloadAbs,
        EnvelopeView env, out string? error)
    {
        error = null;
        if (!TryFollowTable(span, payloadAbs, out var cbFields, out error)) return false;
        if (!TryReadUInt16Field(span, payloadAbs, cbFields, 0, out var schemaVersion, out error))
            return false;
        env.SchemaVersion = schemaVersion;

        if (!TryReadUOffsetFieldLocation(span, payloadAbs, cbFields, 1, out var cmdsStoreAbs,
                out var cmdsRel, out error)) return false;
        var cmdsVecAbs = (int)(cmdsStoreAbs + cmdsRel);
        if (cmdsVecAbs + 4 > span.Length) { error = "commands vector slot past EOF"; return false; }
        var cmdCount = (int)BinaryPrimitives.ReadUInt32LittleEndian(span[cmdsVecAbs..(cmdsVecAbs + 4)]);
        var cmdVecDataAbs = cmdsVecAbs + 4;
        for (int i = 0; i < cmdCount; i++)
        {
            var elemAbs = cmdVecDataAbs + i * 4;
            if (elemAbs + 4 > span.Length) { error = "truncated commands vector"; return false; }
            var cmdRel = BinaryPrimitives.ReadUInt32LittleEndian(span[elemAbs..(elemAbs + 4)]);
            var cmdAbs = elemAbs + (int)cmdRel;
            if (cmdAbs + 4 > span.Length) { error = $"command[{i}] table off {cmdAbs} past EOF"; return false; }
            if (!TryFollowTable(span, cmdAbs, out var cmdFields, out error)) return false;
            if (!TryReadUInt8Field(span, cmdAbs, cmdFields, 7, out var kindByte, out error))
                return false;
            int payloadLen = 0;
            if (HasField(cmdFields, 9))
            {
                if (!TryReadUOffsetFieldLocation(span, cmdAbs, cmdFields, 9, out var plStoreAbs,
                        out var plRel, out error))
                    return false;
                var plAbs = (int)(plStoreAbs + plRel);
                if (!TryReadVector(span, plAbs, out var plCount, out error))
                    return false;
                payloadLen = plCount;
            }
            env.Commands.Add(new CommandView(kindByte, payloadLen));
        }
        return true;
    }

    private static bool DecodeProjectionDelta(ReadOnlySpan<byte> span, int payloadAbs,
        EnvelopeView env, out string? error)
    {
        error = null;
        if (!TryFollowTable(span, payloadAbs, out var pdFields, out error)) return false;

        env.CampaignIdByteCount = ReadByteVectorLen(span, payloadAbs, pdFields, 0, out error);
        if (env.CampaignIdByteCount < 0) return false;
        env.ObserverIdByteCount = ReadByteVectorLen(span, payloadAbs, pdFields, 1, out error);
        if (env.ObserverIdByteCount < 0) return false;
        if (HasField(pdFields, 2))
        {
            var abs = payloadAbs + pdFields[2];
            if (abs + 4 > span.Length) { error = "base_revision past EOF"; return false; }
            env.BaseRevision = BinaryPrimitives.ReadUInt32LittleEndian(span.Slice(abs, 4));
        }
        if (HasField(pdFields, 3))
        {
            var abs = payloadAbs + pdFields[3];
            if (abs + 4 > span.Length) { error = "new_revision past EOF"; return false; }
            env.NewRevision = BinaryPrimitives.ReadUInt32LittleEndian(span.Slice(abs, 4));
        }
        if (HasField(pdFields, 4))
        {
            var abs = payloadAbs + pdFields[4];
            if (abs + 4 > span.Length) { error = "tick past EOF"; return false; }
            env.Tick = BinaryPrimitives.ReadUInt32LittleEndian(span.Slice(abs, 4));
        }
        env.StateHashByteCount = ReadByteVectorLen(span, payloadAbs, pdFields, 5, out error);
        if (env.StateHashByteCount < 0) return false;
        if (!TryReadUOffsetFieldLocation(span, payloadAbs, pdFields, 6, out var decStoreAbs,
                out var decRel, out error)) return false;
        var decVecAbs = (int)(decStoreAbs + decRel);
        if (decVecAbs + 4 > span.Length) { error = "decisions vector slot past EOF"; return false; }
        var decCount = (int)BinaryPrimitives.ReadUInt32LittleEndian(span[decVecAbs..(decVecAbs + 4)]);
        var decVecDataAbs = decVecAbs + 4;
        for (int i = 0; i < decCount; i++)
        {
            var elemAbs = decVecDataAbs + i * 4;
            if (elemAbs + 4 > span.Length) { error = "truncated decisions vector"; return false; }
            var dRel = BinaryPrimitives.ReadUInt32LittleEndian(span[elemAbs..(elemAbs + 4)]);
            var dAbs = elemAbs + (int)dRel;
            if (dAbs + 4 > span.Length) { error = $"decision[{i}] table off {dAbs} past EOF"; return false; }
            if (!TryFollowTable(span, dAbs, out var dFields, out error)) return false;
            ushort codeByte = 0;
            if (HasField(dFields, 2))
            {
                var abs = dAbs + dFields[2];
                if (abs + 2 > span.Length) { error = "decision code past EOF"; return false; }
                codeByte = BinaryPrimitives.ReadUInt16LittleEndian(span.Slice(abs, 2));
            }
            int reasonKeyLen = 0;
            if (HasField(dFields, 3))
            {
                if (!TryReadUOffsetFieldLocation(span, dAbs, dFields, 3, out var rkStoreAbs,
                        out var rkRel, out error)) return false;
                var rkAbs = (int)(rkStoreAbs + rkRel);
                if (!TryReadVector(span, rkAbs, out var rkCount, out error))
                    return false;
                reasonKeyLen = rkCount;
            }
            int detailsLen = 0;
            if (HasField(dFields, 5))
            {
                if (!TryReadUOffsetFieldLocation(span, dAbs, dFields, 5, out var dtStoreAbs,
                        out var dtRel, out error)) return false;
                var dtAbs = (int)(dtStoreAbs + dtRel);
                if (!TryReadVector(span, dtAbs, out var dtCount, out error))
                    return false;
                detailsLen = dtCount;
            }
            env.Decisions.Add(new DecisionView(codeByte, reasonKeyLen, detailsLen));
        }
        env.ViewsByteCount = ReadByteVectorLen(span, payloadAbs, pdFields, 7, out error);
        if (env.ViewsByteCount < 0) return false;
        env.RemovalsByteCount = ReadByteVectorLen(span, payloadAbs, pdFields, 8, out error);
        if (env.RemovalsByteCount < 0) return false;
        env.AlertsByteCount = ReadByteVectorLen(span, payloadAbs, pdFields, 9, out error);
        if (env.AlertsByteCount < 0) return false;
        env.ExplanationsByteCount = ReadByteVectorLen(span, payloadAbs, pdFields, 10, out error);
        if (env.ExplanationsByteCount < 0) return false;
        return true;
    }

    private static int ReadByteVectorLen(ReadOnlySpan<byte> span, int tableOff, int[] fields, int idx,
        out string? error)
    {
        error = null;
        if (!HasField(fields, idx)) return 0;
        if (!TryReadUOffsetFieldLocation(span, tableOff, fields, idx, out var storeAbs, out var rel, out error))
            return -1;
        var vecAbs = (int)(storeAbs + rel);
        if (!TryReadVector(span, vecAbs, out var count, out error))
            return -1;
        return count;
    }

    private static bool TryFollowTable(ReadOnlySpan<byte> buf, int tableOff, out int[] fields, out string error)
    {
        fields = Array.Empty<int>();
        error = "";
        if (tableOff < 0 || tableOff + 4 > buf.Length)
        { error = $"table off {tableOff} past EOF"; return false; }
        var soff = BinaryPrimitives.ReadInt32LittleEndian(buf.Slice(tableOff, 4));
        var vtableOff = tableOff - soff;
        if (soff == 0 || vtableOff < 0 || vtableOff + 4 > buf.Length)
        { error = $"vtable soffset {soff} invalid for table at {tableOff}"; return false; }
        var vtableSize = BinaryPrimitives.ReadUInt16LittleEndian(buf.Slice(vtableOff, 2));
        if (vtableOff + vtableSize > buf.Length)
        { error = "vtable truncated"; return false; }
        var fieldCount = (vtableSize - 4) / 2;
        fields = new int[fieldCount];
        for (int i = 0; i < fieldCount; i++)
            fields[i] = BinaryPrimitives.ReadUInt16LittleEndian(buf.Slice(vtableOff + 4 + 2 * i, 2));
        return true;
    }

    private static bool HasField(int[] fields, int idx) => idx < fields.Length && fields[idx] != 0;

    private static bool TryReadUInt8Field(ReadOnlySpan<byte> buf, int tableOff, int[] fields, int idx,
        out byte value, out string error)
    {
        value = 0; error = "";
        if (!HasField(fields, idx)) { value = 0; return true; }
        var abs = tableOff + fields[idx];
        if (abs + 1 > buf.Length) { error = "u8 field past EOF"; return false; }
        value = buf[abs];
        return true;
    }

    private static bool TryReadUInt16Field(ReadOnlySpan<byte> buf, int tableOff, int[] fields, int idx,
        out ushort value, out string error)
    {
        value = 0; error = "";
        if (!HasField(fields, idx)) { value = 0; return true; }
        var abs = tableOff + fields[idx];
        if (abs + 2 > buf.Length) { error = "u16 field past EOF"; return false; }
        value = BinaryPrimitives.ReadUInt16LittleEndian(buf.Slice(abs, 2));
        return true;
    }

    private static bool TryReadUOffsetFieldLocation(ReadOnlySpan<byte> buf, int tableOff, int[] fields,
        int idx, out uint storeAbs, out uint rel, out string error)
    {
        storeAbs = 0; rel = 0; error = "";
        if (!HasField(fields, idx)) return true;
        var abs = tableOff + fields[idx];
        if (abs + 4 > buf.Length) { error = "uoffset field past EOF"; return false; }
        storeAbs = (uint)abs;
        rel = BinaryPrimitives.ReadUInt32LittleEndian(buf.Slice(abs, 4));
        return true;
    }

    private static bool TryReadVector(ReadOnlySpan<byte> span, int vecAbs, out int count, out string? error)
    {
        count = 0; error = null;
        if (vecAbs + 4 > span.Length) { error = "vector slot past EOF"; return false; }
        var rel = BinaryPrimitives.ReadUInt32LittleEndian(span[vecAbs..(vecAbs + 4)]);
        var dataAbs = vecAbs + (int)rel;
        if (dataAbs + 4 > span.Length) { error = "vector count past EOF"; return false; }
        count = (int)BinaryPrimitives.ReadUInt32LittleEndian(span[dataAbs..(dataAbs + 4)]);
        return true;
    }
}

File: tests/flatbuffers_reader.py
"""Hand-rolled FlatBuffers reader used by the WP02-B cross-language round-trip gate.

Supports only what the audit needs:
  - 4-byte file_identifier lock (CSWP)
  - root uoffset (must be safe to follow)
  - Envelope: payload_type (uint8), payload (union table offset)
  - CommandBatch: schema_version (uint32), commands (vector of tables)
  - CommandEnvelope: kind (uint16), priority (int32), submitted_tick (uint64),
    campaign_id/issuer_id/payload (vectors of uint8), command_id ([uint8;16])

Mirrors the wire layout documented in Google FlatBuffers (Apache 2.0) without
pulling in the Python library. Bounds-checks every offset to reject corruption,
truncation, and arbitrary-length attacks.

The reader is cross-validated against `flatc --binary` output via
`flatc --json --raw-binary --strict-json` against the committed golden fixture
at .agileplus/civic-warfare-program/contracts/fixtures/sample-envelope.bin.

CLI:
  python3 flatbuffers_reader.py decode <fixture.bin>     # pretty-print decoded
  python3 flatbuffers_reader.py schema-version <fixture.bin>   # print schema_version
  python3 flatbuffers_reader.py commands-count <fixture.bin>    # print commands count
  python3 flatbuffers_reader.py first-command-kind <fixture.bin> # print first command kind
"""

from __future__ import annotations

import json
import struct
import sys
from pathlib import Path

# ---------------------------------------------------------------------------
# Authoritative schema (must match .agileplus/civic-warfare-program/contracts/warfare.fbs)
# ---------------------------------------------------------------------------

# enum CommandKind : ushort (offset-by-zero in declaration order)
COMMAND_KIND_VALUES = (
    "None",
    "RaiseAlert",
    "ReduceAlert",
    "MobilizeReserve",
    "DemobilizeReserve",
    "RequisitionEquipment",
    "RequisitionPersonnel",
    "AllocateResources",
    "TransferFunds",
    "EnterNegotiation",
    "ExecuteMission",
    "IssueDirective",
    "DispatchAdvisory",
    "RecallUnit",
)
COMMAND_KIND_INDEX = {name: idx for idx, name in enumerate(COMMAND_KIND_VALUES)}

# enum RootPayload : byte
ROOT_PAYLOAD_VALUES = ("NONE", "CommandBatch", "ProjectionDelta", "SaveEnvelope")
ROOT_PAYLOAD_INDEX = {name: idx for idx, name in enumerate(ROOT_PAYLOAD_VALUES)}

EXPECTED_FILE_IDENTIFIER = b"CSWP"


# ---------------------------------------------------------------------------
# Low-level readers (bounds-checked)
# ---------------------------------------------------------------------------


class FlatbuffersError(Exception):
    """Raised when the input buffer is malformed, truncated, or contains
    illegal FlatBuffers offsets (corruption, malicious input, etc.).
    """


def _need(buf: bytes, off: int, n: int, what: str) -> None:
    if off < 0 or off + n > len(buf):
        raise FlatbuffersError(f"{what} out of bounds: {off}+{n} > {len(buf)}")


def _u8(buf: bytes, off: int) -> int:
    _need(buf, off, 1, "u8")
    return buf[off]


def _u16(buf: bytes, off: int) -> int:
    _need(buf, off, 2, "u16")
    return struct.unpack_from("<H", buf, off)[0]


def _u32(buf: bytes, off: int) -> int:
    _need(buf, off, 4, "u32")
    return struct.unpack_from("<I", buf, off)[0]


def _u64(buf: bytes, off: int) -> int:
    _need(buf, off, 8, "u64")
    return struct.unpack_from("<Q", buf, off)[0]


def _i32(buf: bytes, off: int) -> int:
    _need(buf, off, 4, "i32")
    return struct.unpack_from("<i", buf, off)[0]


def _follow_table(buf: bytes, table_off: int) -> tuple[int, int, list[int]]:
    """Follow the vtable from the table position. Returns
    (vtable_off, table_inline_size, field_voffs).

    Raises FlatbuffersError on malformed input.
    """
    if table_off < 4 or table_off + 4 > len(buf):
        raise FlatbuffersError(f"table position out of bounds: {table_off}")

    # First int32 at the table is the signed offset to the vtable.
    # Per FlatBuffers spec, vtable_off = table_off + vsoff (can be negative,
    # Per FlatBuffers spec: vtable_off = table_off - vsoff (signed int32).
    # soff is stored as the unsigned difference (table - vtable); a typical
    # value is +8 (vtable sits 8 bytes before the table). With signed
    # subtraction that yields table_off - 8 = correct vtable position.
    vsoff = _i32(buf, table_off)
    vtable_off = table_off - vsoff
    if vtable_off < 0 or vtable_off + 4 > len(buf):
        raise FlatbuffersError(
            f"vtable follow: table_off={table_off}, vsoff={vsoff}, "
            f"vtable_off={vtable_off} out of bounds (len={len(buf)})"
        )
    vsize = _u16(buf, vtable_off)
    if vsize < 4 or vtable_off + vsize > len(buf):
        raise FlatbuffersError(f"vtable size out of bounds: {vsize}")
    tsize = _u16(buf, vtable_off + 2)
    field_count = (vsize - 4) // 2

    # Validate field offsets
    field_voffs: list[int] = []
    for i in range(field_count):
        voff = _u16(buf, vtable_off + 4 + 2 * i)
        if voff != 0:
            # Inlined scalar field at table_off + voff
            # (or uoffset field: target = table_off + voff + read_u32(target))
            # We don't validate the uoffset target here; reader does that.
            if voff > tsize:
                raise FlatbuffersError(
                    f"field[{i}] voff {voff} exceeds table inline size {tsize}"
                )
        field_voffs.append(voff)

    return vtable_off, tsize, field_voffs


def _follow_uoffset(buf: bytes, slot_off: int) -> int:
    """Follow a uoffset field: target = slot_off + rel (rel is read at slot_off)."""
    rel = _u32(buf, slot_off)
    target = slot_off + rel
    if target < 0 or target >= len(buf):
        raise FlatbuffersError(
            f"uoffset follow: slot={slot_off}, rel={rel}, target={target} oob"
        )
    return target


# ---------------------------------------------------------------------------
# High-level decoders (per FlatBuffers table)
# ---------------------------------------------------------------------------


def decode_envelope(buf: bytes) -> dict:
    """Decode the Envelope root table. Returns a JSON-serializable dict."""
    if len(buf) < 8:
        raise FlatbuffersError(f"buffer too short for Envelope header: {len(buf)}")

    # File identifier lock (bytes 4..8)
    if buf[4:8] != EXPECTED_FILE_IDENTIFIER:
        raise FlatbuffersError(
            f"file_identifier mismatch: got {buf[4:8]!r}, "
            f"expected {EXPECTED_FILE_IDENTIFIER!r}"
        )

    # Root uoffset (bytes 0..4)
    root_off = _u32(buf, 0)
    if root_off >= len(buf):
        raise FlatbuffersError(f"root uoffset out of bounds: {root_off}")

    _, _, env_fields = _follow_table(buf, root_off)

    # Envelope fields:
    # 0: payload_type (uint8) - inlined scalar
    # 1: payload (uoffset to CommandBatch/ProjectionDelta/SaveEnvelope table)

    payload_type = (
        _u8(buf, root_off + env_fields[0]) if len(env_fields) > 0 and env_fields[0] else 0
    )
    payload_type_name = (
        ROOT_PAYLOAD_VALUES[payload_type]
        if payload_type < len(ROOT_PAYLOAD_VALUES)
        else f"Unknown({payload_type})"
    )

    payload = None
    if len(env_fields) > 1 and env_fields[1]:
        payload_slot = root_off + env_fields[1]
        payload_off = _follow_uoffset(buf, payload_slot)
        payload = _decode_payload(buf, payload_off, payload_type_name)

    return {
        "file_identifier": buf[4:8].decode("ascii", errors="replace"),
        "root_off": root_off,
        "payload_type": payload_type_name,
        "payload": payload,
    }


def _decode_payload(buf: bytes, payload_off: int, payload_type_name: str) -> dict:
    if payload_type_name == "CommandBatch":
        return _decode_command_batch(buf, payload_off)
    # Other union variants: only CommandBatch is required for the audit gate.
    raise FlatbuffersError(f"unsupported payload type: {payload_type_name}")


def _decode_command_batch(buf: bytes, cb_off: int) -> dict:
    _, _, cb_fields = _follow_table(buf, cb_off)

    # CommandBatch fields:
    # 0: schema_version (uint32)
    # 1: commands (uoffset to vector of CommandEnvelope)
    schema_version = (
        _u32(buf, cb_off + cb_fields[0])
        if len(cb_fields) > 0 and cb_fields[0]
        else 0
    )

    commands: list[dict] = []
    if len(cb_fields) > 1 and cb_fields[1]:
        cmds_slot = cb_off + cb_fields[1]
        cmds_off = _follow_uoffset(buf, cmds_slot)
        cmds_count = _u32(buf, cmds_off)
        cmds_data = cmds_off + 4
        for k in range(cmds_count):
            elem_slot = cmds_data + k * 4
            cmd_off = _follow_uoffset(buf, elem_slot)
            commands.append(_decode_command_envelope(buf, cmd_off))

    return {
        "schema_version": schema_version,
        "commands": commands,
    }


def _decode_command_envelope(buf: bytes, cmd_off: int) -> dict:
    _, _, f = _follow_table(buf, cmd_off)

    # CommandEnvelope fields (from warfare.fbs):
    # 0: command_id ([uint8; 16])  - inline fixed-size array
    # 1: campaign_id (vector<uint8>)
    # 2: issuer_id (vector<uint8>)
    # 3: submitted_tick (uint64)
    # 4: scheduled_tick (uint64)
    # 5: priority (int32)
    # 6: expected_revision (int32)
    # 7: kind (ushort)  - CommandKind enum
    # 8: payload (vector<uint8>)
    # 9: notes (string)

    out: dict = {}

    # command_id: 16-byte inline fixed array at cmd_off + 4 (after vtable soff)
    # But per FlatBuffers, inline arrays are stored inline at cmd_off + voff[0].
    if len(f) > 0 and f[0]:
        cid_off = cmd_off + f[0]
        _need(buf, cid_off, 16, "command_id")
        out["command_id"] = list(buf[cid_off : cid_off + 16])

    # vectors
    for vec_idx, name in [(1, "campaign_id"), (2, "issuer_id"), (8, "payload")]:
        if len(f) > vec_idx and f[vec_idx]:
            vslot = cmd_off + f[vec_idx]
            voff = _follow_uoffset(buf, vslot)
            vcount = _u32(buf, voff)
            vdata = voff + 4
            _need(buf, vdata, vcount, f"{name} data")
            out[name] = list(buf[vdata : vdata + vcount])

    # uint64
    if len(f) > 3 and f[3]:
        out["submitted_tick"] = _u64(buf, cmd_off + f[3])
    if len(f) > 4 and f[4]:
        out["scheduled_tick"] = _u64(buf, cmd_off + f[4])

    # int32
    if len(f) > 5 and f[5]:
        out["priority"] = _i32(buf, cmd_off + f[5])
    if len(f) > 6 and f[6]:
        out["expected_revision"] = _i32(buf, cmd_off + f[6])

    # ushort enum
    if len(f) > 7 and f[7]:
        kind_raw = _u16(buf, cmd_off + f[7])
        out["kind"] = (
            COMMAND_KIND_VALUES[kind_raw]
            if kind_raw < len(COMMAND_KIND_VALUES)
            else f"Unknown({kind_raw})"
        )

    # string (notes): uoffset to vector<uint8> + null terminator expected
    if len(f) > 9 and f[9]:
        nslot = cmd_off + f[9]
        noff = _follow_uoffset(buf, nslot)
        ncount = _u32(buf, noff)
        ndata = noff + 4
        _need(buf, ndata, ncount, "notes data")
        # Strip trailing NUL (FlatBuffers string convention)
        raw = buf[ndata : ndata + ncount]
        if raw and raw[-1] == 0:
            raw = raw[:-1]
        out["notes"] = raw.decode("utf-8", errors="replace")

    return out


# ---------------------------------------------------------------------------
# CLI
# ---------------------------------------------------------------------------


def _cli_decode(path: Path) -> int:
    obj = decode_envelope(path.read_bytes())
    print(json.dumps(obj, indent=2, sort_keys=True))
    return 0


def _cli_scalar(path: Path, key: str) -> int:
    obj = decode_envelope(path.read_bytes())
    payload = obj.get("payload") or {}
    if key == "schema-version":
        print(payload.get("schema_version"))
    elif key == "commands-count":
        print(len(payload.get("commands") or []))
    elif key == "first-command-kind":
        cmds = payload.get("commands") or []
        print(cmds[0].get("kind") if cmds else "<none>")
    elif key == "second-command-kind":
        cmds = payload.get("commands") or []
        print(cmds[1].get("kind") if len(cmds) > 1 else "<none>")
    else:
        raise SystemExit(f"unknown scalar key: {key}")
    return 0


def main(argv: list[str]) -> int:
    if len(argv) < 3:
        print(__doc__, file=sys.stderr)
        return 2
    cmd = argv[1]
    path = Path(argv[2])
    if cmd == "decode":
        return _cli_decode(path)
    if cmd in ("schema-version", "commands-count", "first-command-kind", "second-command-kind"):
        return _cli_scalar(path, cmd)
    print(f"unknown command: {cmd}", file=sys.stderr)
    return 2


if __name__ == "__main__":
    sys.exit(main(sys.argv))
